### DISCUSSION TITLE
Potential solution and causes to shadcn/radix component issues

### DISCUSSION THREAD

**Discussion by pact0** [wxt-dev/wxt#1300] [General]
Hello, I've been rewriting my extension using a bit more modern stack and I ended up using this framework and tailwind + shadcn components. I've almost instantly noticed that some of the styles work and some don't for shadcn. I've spent some time debugging it and found out why so I might as well share. 

This seems to be an issue only in the content scripts. Some of the simple components like button or label work just fine, but dropdown doesn't.

Here is a reproduction repo which is some random starter with added components https://github.com/pact0/reproduction . When you look at the sidepanel entrypoint it uses the same dropdown as content script and in sidepanel the component is styled correctly in content script it is not. 

The issue is shadcn uses radix undearneath which uses a concept called Portals to render things like dropdowns https://www.radix-ui.com/primitives/docs/utilities/portal in the documentation it says that it appends to the document.body by default and this is the culprit. When injecting the content script we create the shadow root which holds all the css we need. but the portaled components don't have access to them since the live in the root body. 

Here is also a temp solution commented out which makes it work https://github.com/pact0/reproduction/blob/main/src/entrypoints/content/index.tsx just adding our styles to the original website, which is pretty bad, but works.

The ideal solution would  be for shadcn to expose the portal custom container api somehow and allow us to change where these components get appended. I've seen a bunch of people ask about this issue and haven't seen anyone mention why this is happening. I hope this helps to somehow fix this

**Comment by nishu-murmu** [comment]
This problem is same with any other UI component library. Like for example I've grown fond of headlessui components, I use that a lot, this package also has Dropdown like component known as Combobox, when I wanted to configure this in my content scripts, I also faced this same issue.

From what I've debugged and scour through the github issues,
There is one work around
*i.e.* to use MutationObserver and listen to the DOM changes for the portal, and move the portal element in the shadow DOM of the content scripts.

*Edit:*
Apparently they fixed it in headlessui with this [PR](https://github.com/tailwindlabs/headlessui/pull/3594).

**Comment by pact0** [comment]
Would be nice to spread the awareness and have shadcn team fix this issue as well, because currently injecting the styles to the main document works,but messes up original website styles most of the time ;/

**Reply by nishu-murmu** [reply]
Don't inject styles to main document, cause different websites will have different styling, for sure, it will break the styling.
For temporary workaround, please use **MutationObsever**.

And yeah, someone should create a github discussions or issue in the shadcn community.

**Reply by pact0** [reply]
do you maybe have an example how can I move it to the shadow dom?

**Reply by nishu-murmu** [reply]
I tried to make it work with the reproduction repo you sent.
Styling works but behaviour got messed up.

Here is the code snippet

```
export function customMutationObserver({
  callback,
  targetElem,
  options,
}: CustomMutationObserver) {
  const observer = new MutationObserver((mutations) => {
    callback(mutations, observer);
  });
  observer.observe(
    targetElem || document.body,
    options || {
      childList: true,
      subtree: true,
    }
  );
}
```

```
customMutationObserver({
    callback: async (list, observer) => {
     // replace this with the portal element for radix
      const portalRoot = document.querySelector('#headlessui-portal-root');
      // root elem of content scripts injected
      const dropdown = document.querySelector(
        'enacton-crm-assignee-dropdown'
      );

      if (portalRoot) {
        dropdown.shadowRoot?.appendChild(portalRoot);
        portalRoot.remove();
        observer.disconnect();
      }
    },
  });
```

**Comment by aklinker1** [comment]
https://wxt.dev/guide/resources/faq.html#my-component-library-doesn-t-work-in-content-scripts

Very similar to this, but instead of styles being outside the shadow root, elements are being placed outside the shadow root.

Doesn't look like Shadcn documents how to change this, I guess you need to provide your own solution on top of it.... But you say it's not possible? Doesn't Shadcn just download component code, can't you just change the portal container properly?

The simplest thing would be to use react context to provide the element to add portals to.

**Reply by pact0** [reply]
That's what also crossed my mind, I'll try to have a look at this, thanks for the suggestions

**Comment by Maks-T** [comment]
Hi everyone! 👋

If you're dealing with Tailwind CSS class collisions or style leaks when sharing UI components / embedding apps in a monorepo, I built an AST-driven CLI tool that automates Tailwind prefixing and complete style isolation: **[`tailwind-prefix-migrator`](https://github.com/Maks-T/tailwind-prefix-migrator)**.

### Key Features:
- 🎯 **AST-Driven Precision (Babel & PostCSS):** Safely prefixes Tailwind classes across JS, TSX, SCSS, `@apply` rules, and `cva()` definitions without regular expression edge-case bugs.
- 🛡 **CSS Layer & Variable Isolation:** Automatically wraps `@tailwind` directives in custom `@layer` rules and prefixes CSS variables (e.g., `--primary` ➔ `--calc-primary`) to prevent style leaks into host applications.
- 📦 **Config Sync:** Updates `prefix` and `preflight` settings in `tailwind.config.js` and `components.json` (shadcn/ui).
- 🔄 **Instant Rollback:** Reverse prefix migrations anytime using `tpm revert`.

You can test it in dry-run mode via `npx`:

```bash
npx tailwind-prefix-migrator init
npx tailwind-prefix-migrator apply --dry-run
